pkg/generate/code/late_initialization.go

```diff
@@ -0,0 +1,246 @@
+	}
+	out += fmt.Sprintf("}\n")
+	return out
+}
+
+func lateInitMap(responsePath, crPath string, r *model.CRD, str *awssdkmodel.ShapeRef, level int) string {
+	out := fmt.Sprintf("if %s != nil {\n", responsePath)
+	out += fmt.Sprintf("if %s == nil {\n", crPath)
+	out += fmt.Sprintf("%s = %s{}\n", crPath, GetCorrespondingCRDType(str.Shape, r, false))
+	out += fmt.Sprintf("}\n")
+
+	out += fmt.Sprintf("for key%d := range %s {\n", level, responsePath)
+	respFieldPath := fmt.Sprintf("%s[key%d]", responsePath, level)
+	crFieldPath := fmt.Sprintf("%s[key%d]", crPath, level)
+	out += lateInit(respFieldPath, crFieldPath, r, str, level+1)
+	out += fmt.Sprintf("}\n")
+	out += fmt.Sprintf("}\n")
+	return out
+}
+
+func lateInitSlice(responsePath, crPath string, r *model.CRD, respShapeRef *awssdkmodel.ShapeRef, level int) string {
+	// NOTE(muvaf): Late initialization for slices is not a full-featured late-init.
+	// It works only if the slice in CRD is empty. If there is even one element,
+	// we don't late-init since slices do not have unique identifier like map/struct.
+	// We're playing conservative and not rely on index so that we don't risk
+	// overriding desired state with what's observed.
+	// TODO(muvaf): Kubernetes Apply uses comment markers on fields to specify an
+	// identifier (`name` most of the time). If we can do such thing, we can implement
+	// full late-init for slice elements as well, just like maps.
+	out := fmt.Sprintf(
+		"if len(%s) != 0 && len(%s) == 0 {\n", responsePath, crPath,
+	)
+	out += fmt.Sprintf("%s = make([]%s, len(%s))\n", crPath, GetCorrespondingCRDType(respShapeRef.Shape.MemberRef.Shape, r, true), responsePath)
+	out += fmt.Sprintf("for i%d := range %s {\n", level, responsePath)
+	respFieldPath := fmt.Sprintf("%s[i%d]", responsePath, level)
+	crFieldPath := fmt.Sprintf("%s[i%d]", crPath, level)
+	out += lateInit(respFieldPath, crFieldPath, r, &respShapeRef.Shape.MemberRef, level+1)
+	out += fmt.Sprintf("}\n")
+	out += fmt.Sprintf("}\n")
+	return out
+}
+
+func lateInit(responsePath, crPath string, r *model.CRD, str *awssdkmodel.ShapeRef, level int) string {
+	switch str.Shape.Type {
+	case "string":
+		return fmt.Sprintf("%s = awsclients.LateInitializeStringPtr(%s, %s)\n", crPath, crPath, responsePath)
+	case "long", "integer":
+		return fmt.Sprintf("%s = awsclients.LateInitializeInt64Ptr(%s, %s)\n", crPath, crPath, responsePath)
+	case "boolean":
+		return fmt.Sprintf("%s = awsclients.LateInitializeBoolPtr(%s, %s)\n", crPath, crPath, responsePath)
+	case "list":
+		return lateInitSlice(responsePath, crPath, r, str, level)
+	case "structure":
+		return lateInitStruct(responsePath, crPath, r, str, level)
+	case "map":
+		return lateInitMap(responsePath, crPath, r, str, level)
+	default:
+		panic(fmt.Sprintf("unknown shape type %s", str.Shape.Type))
+	}
+}
+
+func GetCorrespondingCRDType(s *awssdkmodel.Shape, r *model.CRD, keepPointer bool) string {
```

Comment:
Definitely this needs a docstring :) Also, let's rename this function to be more explicit. *What* specifically is the returned type "corresponding" to? Perhaps rename to `GetCRDStructType()` to match the aws-sdk-go naming convention?